### PR TITLE
Feature/ecs provider

### PR DESCRIPTION
**Pull request by ycliuhw**
*Wireframe the ECS provider;*
Implemented features:

- Ensure() - deploy the task definition and service;
- Delete() - remove the whole stack when the application gets deleted;

## Checklist

 - [ ] ~Requires a [pylibjuju](https://github.com/juju/python-libjuju) change~
 - [ ] ~Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR~
 - [ ] ~Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed~
 - [x] Comments answer the question of why design decisions were made

## QA steps

No QA steps, because the ECS integration is not complete yet.

## Documentation changes

No

## Bug reference

No

### CODE REVIEW

File: agent/agentbootstrap/bootstrap.go
Comment (achilleasa) [inline agent/agentbootstrap/bootstrap.go:391]: Could we make this a method on `Type` instead?

File: caas/ecs/application.go
Comment (achilleasa) [inline caas/ecs/application.go:98]: Are we using host-based volumes here? If so, how are we going to deal with the possibility that two task instances of the same app get scheduled on the same node?

File: caas/ecs/application.go
Comment (achilleasa) [inline caas/ecs/application.go:209]: So does this actually work as-is when you use the API? Does the SDK transparently encode the KV pairs when calling to the ECS API? I had issues when using the cli with a json document.

File: caas/ecs/application.go
Comment (achilleasa) [inline caas/ecs/application.go:301]: Are we going for EBS instead of instance storage? Is this so we can always re-attach storage (similar to k8s) when the tasks get rescheduled to other containers?

File: caas/ecs/application.go
Comment (ycliuhw) [inline caas/ecs/application.go:135]: These volumes are configured to `shared` for all the tasks for now.
We will need to test how to set here to `task` scope, so each task has its own volume.

File: caas/ecs/application.go
Comment (ycliuhw) [reply caas/ecs/application.go:110]: We don't use any host-based volumes.
For juju related volumes, we use `DockerVolumeConfiguration` with local driver 

- https://docs.aws.amazon.com/AmazonECS/latest/developerguide/docker-volumes.html.
- https://github.com/juju/juju/pull/12406/files#diff-2ccee37d2965d4d7d708a406d87294f7be801db49dde62eec02686efbbee5709R257

For user-requested volumes, we use `EBS` 
- https://github.com/juju/juju/pull/12406/files#diff-2ccee37d2965d4d7d708a406d87294f7be801db49dde62eec02686efbbee5709R132

File: caas/ecs/application.go
Comment (ycliuhw) [reply caas/ecs/application.go:233]: It works as expected.

File: caas/ecs/application.go
Comment (ycliuhw) [reply caas/ecs/application.go:326]: Yes, We choose this as the default driver for EBS support.
- https://github.com/rexray/rexray

File: caas/ecs/application.go
Comment (ycliuhw) [reply caas/ecs/application.go:110]: We don't use any host-based volumes.
For juju related volumes, we use `DockerVolumeConfiguration` with local driver 

- https://docs.aws.amazon.com/AmazonECS/latest/developerguide/docker-volumes.html.
- https://github.com/juju/juju/pull/12406/files#diff-2ccee37d2965d4d7d708a406d87294f7be801db49dde62eec02686efbbee5709R257

For user-requested volumes, we use `EBS` 
- https://github.com/juju/juju/pull/12406/files#diff-2ccee37d2965d4d7d708a406d87294f7be801db49dde62eec02686efbbee5709R132

File: caas/ecs/application.go
Comment (ycliuhw) [reply caas/ecs/application.go:233]: It works as expected.

Comment (achilleasa) [reply caas/ecs/application.go:233]: Nice!

File: caas/ecs/application.go
Comment (achilleasa) [reply caas/ecs/application.go:135]: According to the docs: ```The scope for the Docker volume, which determines its lifecycle. Docker volumes that are scoped to a task are automatically provisioned when the task starts destroyed when the task is cleaned up. Docker volumes that are scoped as shared persist after the task stops.```. 

Does this mean that the storage gets reclaimed when a task gets moved to another node or clean up only applies when the task is _permanently_ destroyed (e.g. we scale down)?

Comment (ycliuhw) [reply caas/ecs/application.go:135]: The ECS tasks are never restarted but always recreated if any `essential` containers died.
So it seems there is no point to keep the volume after the task died because there will be no chance to be re-attached again later at some point.
So we will need to test what happens it sets to the `task` scope.

File: agent/agentbootstrap/bootstrap.go
Comment (ycliuhw) [reply agent/agentbootstrap/bootstrap.go:391]: It is a good idea.
I tried to move this to cloud.Type. But I found I have to touch hundreds of files (from `cmd` to `api/apiserver` then `state`).
Given this PR is already not small. It might be better to do this change in a separate PR.

**Comment by ycliuhw** [comment]
[Add a new cloud type for ecs(might need tweak add-cloud cmd);](https://trello.com/c/QVXwRTsQ/1308-add-a-new-cloud-type-for-ecsmight-need-tweak-add-cloud-cmd)

**Comment by ycliuhw** [comment]
[Implement ECS API integration under `caas/ecs` package](https://trello.com/c/XfUkGxAk/1309-implement-ecs-api-integration-under-caas-ecs-package)

File: worker/caasapplicationprovisioner/application.go
Comment (achilleasa) [inline worker/caasapplicationprovisioner/application.go:199]: The prior behavior (mute errors) before your change looks quite dangerous... Should the changes in this method be backported to 2.9?

**Review by achilleasa — APPROVED** [review]
Overall it LGTM. Left some comments (there is a missing insert that needs to be fixed). I will go ahead and approve it so you can land it after you resolve the broken tests and apply any of my comments that make sense.

I will iterate on the provider once this patch lands and start populating the TODOs and try to figure out a way to grab the task ID via the ec2 metadata route as we discussed.

File: caas/ecs/constants/constants.go
Comment (ycliuhw) [reply caas/ecs/constants/constants.go:4]: To avoid importing `ecs` package when we need those constants from outside (workers for example).
If we can ensure no need to import these `StorageProviderType` etc constants later, then we can move those constants to `ecs` package.

File: worker/caasapplicationprovisioner/application.go
Comment (ycliuhw) [reply worker/caasapplicationprovisioner/application.go:199]: yeah, I will backport these fixes to 2.9

File: caas/ecs/application.go
Comment (ycliuhw) [reply caas/ecs/application.go:127]: ops, fixing that!

File: caas/ecs/application.go
Comment (ycliuhw) [reply caas/ecs/application.go:497]: Luckily, AWS provides all those utils like `aws.TimeValue`

File: caas/ecs/application.go
Comment (ycliuhw) [reply caas/ecs/application.go:499]: yeah, we can set it to `ExecutionStoppedAt` for now.
We can change when we are able to get the real status later.

**Comment by ycliuhw** [comment]
$$merge$$

File: caas/ecs/package_test.go
Comment (manadart) [inline caas/ecs/package_test.go:4]: Is there an import loop that prevents us putting this in the _ecs_ package and using exported suite methods instead of _export_test.go_

Comment (achilleasa) [reply caas/ecs/package_test.go:4]: We will moving stuff around in the following PRs so we can try changing this in one of those.

**Pull request merged**
